Add initial support for WebGL2 uniform buffer functions

mmatyas opened the pull request:
This *work-in-progress* patch adds initial support for the following WebGL2 calls:

- `bindBufferBase`
- `bindBufferRange`
- `getUniformIndices`
- `getUniformBlockIndex`
- `getActiveUniforms`
- `getActiveUniformBlockParameter`
- `getActiveUniformBlockName`
- `uniformBlockBinding`

See: https://www.khronos.org/registry/webgl/specs/latest/2.0/#3.7.16

However, indexed uniforms and `getIndexedParameter`, which would be used by some of the functions (and transform feedback) is still missing. Also this patch depends on:

- https://github.com/servo/sparkle/pull/16
- https://github.com/servo/servo/issues/25034

(required for both building and running the tests).

cc @jdm @zakorgy @imiklos 

<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

<!-- Either: -->
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

mmatyas committed:
Add initial support for WebGL2 uniform buffer functions

Adds initial support for the following WebGL2 calls:

- bindBufferBase
- bindBufferRange
- getUniformIndices
- getUniformBlockIndex
- getActiveUniforms
- getActiveUniformBlockParameter
- getActiveUniformBlockName
- uniformBlockBinding

See: https://www.khronos.org/registry/webgl/specs/latest/2.0/#3.7.16

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/webidls/WebGL2RenderingContext.webidl, components/script/dom/webglshader.rs, components/script/dom/bindings/trace.rs, components/script/dom/webgl2renderingcontext.rs, components/script/dom/webglprogram.rs
 * @jgraham: tests/wpt/webgl/meta/conformance2/buffers/uniform-buffers.html.ini
 * @kichjang: components/script/dom/webidls/WebGL2RenderingContext.webidl, components/script/dom/webglshader.rs, components/script/dom/bindings/trace.rs, components/script/dom/webgl2renderingcontext.rs, components/script/dom/webglprogram.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

jdm reviewed (approved):
These changes look good!

bors-servo commented:
:umbrella: The latest upstream changes (presumably #25344) made this pull request unmergeable. Please resolve the merge conflicts.

mmatyas commented:
Ok, both the Sparkle side and #25344 seems to have landed, so I've rebased the patch to master and updated the test result.

With these new functions some test can now run further, but produce errors unrelated to the new features (eg. because to `gl.getIndexedParameter` not implemented yet or some parameter is not supported by an unrelated function). These could be improved in follow up PRs.

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit e8e7af6 has been approved by `jdm`

<!-- @bors-servo r=jdm e8e7af6b9d3c57a3f2d2f0da64341acfee9e05dc -->

bors-servo commented:
:hourglass: Testing commit e8e7af6b9d3c57a3f2d2f0da64341acfee9e05dc with merge a57645c2b4c6e56cadf85327b42054d0f3888b3f...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/cCZ7iwrgQO-KyoGcqt906w)

jdm commented:
```
  ▶ Unexpected subtest result in /_webgl/conformance/state/gl-enum-tests.html:
  │ FAIL [expected PASS] WebGL test #20: getError expected: INVALID_ENUM. Was NO_ERROR : gl.bufferData(gl.ARRAY_BUFFER, 16, desktopGL['STREAM_READ']) should return INVALID_ENUM.
  │   → assert_true: getError expected: INVALID_ENUM. Was NO_ERROR : gl.bufferData(gl.ARRAY_BUFFER, 16, desktopGL['STREAM_READ']) should return INVALID_ENUM. expected true got false
  │ FAIL [expected PASS] WebGL test #21: getError expected: INVALID_ENUM. Was NO_ERROR : gl.bufferData(gl.ARRAY_BUFFER, 16, desktopGL['STREAM_COPY']) should return INVALID_ENUM.
  │   → assert_true: getError expected: INVALID_ENUM. Was NO_ERROR : gl.bufferData(gl.ARRAY_BUFFER, 16, desktopGL['STREAM_COPY']) should return INVALID_ENUM. expected true got false
  │ FAIL [expected PASS] WebGL test #22: getError expected: INVALID_ENUM. Was NO_ERROR : gl.bufferData(gl.ARRAY_BUFFER, 16, desktopGL['STATIC_READ']) should return INVALID_ENUM.
  │   → assert_true: getError expected: INVALID_ENUM. Was NO_ERROR : gl.bufferData(gl.ARRAY_BUFFER, 16, desktopGL['STATIC_READ']) should return INVALID_ENUM. expected true got false
  │ FAIL [expected PASS] WebGL test #23: getError expected: INVALID_ENUM. Was NO_ERROR : gl.bufferData(gl.ARRAY_BUFFER, 16, desktopGL['STATIC_COPY']) should return INVALID_ENUM.
  │   → assert_true: getError expected: INVALID_ENUM. Was NO_ERROR : gl.bufferData(gl.ARRAY_BUFFER, 16, desktopGL['STATIC_COPY']) should return INVALID_ENUM. expected true got false
  │ FAIL [expected PASS] WebGL test #24: getError expected: INVALID_ENUM. Was NO_ERROR : gl.bufferData(gl.ARRAY_BUFFER, 16, desktopGL['DYNAMIC_READ']) should return INVALID_ENUM.
  │   → assert_true: getError expected: INVALID_ENUM. Was NO_ERROR : gl.bufferData(gl.ARRAY_BUFFER, 16, desktopGL['DYNAMIC_READ']) should return INVALID_ENUM. expected true got false
  │ FAIL [expected PASS] WebGL test #25: getError expected: INVALID_ENUM. Was NO_ERROR : gl.bufferData(gl.ARRAY_BUFFER, 16, desktopGL['DYNAMIC_COPY']) should return INVALID_ENUM.
  │   → assert_true: getError expected: INVALID_ENUM. Was NO_ERROR : gl.bufferData(gl.ARRAY_BUFFER, 16, desktopGL['DYNAMIC_COPY']) should return INVALID_ENUM. expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/_webgl/js/js-test-pre.js:123:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1984:25
  │ test@http://web-platform.test:8000/resources/testharness.js:548:30
  │ reportTestResultsToHarness@http://web-platform.test:8000/_webgl/js/js-test-pre.js:122:7
  │ testFailed@http://web-platform.test:8000/_webgl/js/js-test-pre.js:249:5
  │ glErrorShouldBeImpl@http://web-platform.test:8000/_webgl/js/webgl-test-utils.js:1679:5
  │ glErrorShouldBe@http://web-platform.test:8000/_webgl/js/webgl-test-utils.js:1653:10
  └ @http://web-platform.test:8000/_webgl/js/tests/gl-enum-tests.js:98:9
```

jdm commented:
I _suspect_ those failures are symptoms of https://github.com/servo/servo/issues/24537.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #25451) made this pull request unmergeable. Please resolve the merge conflicts.

mmatyas commented:
Ok, rebased and fixed by adding usage validation.

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit da94f8d has been approved by `jdm`

<!-- @bors-servo r=jdm da94f8d0e78546a40972596ad0ae4ded510057de -->

bors-servo commented:
:hourglass: Testing commit da94f8d0e78546a40972596ad0ae4ded510057de with merge 123d43bb16cd6807ad6de3e83316191a58505d5b...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/U8pgUbuuSgSVUa_JwJPmcw)
Approved by: jdm
Pushing 123d43bb16cd6807ad6de3e83316191a58505d5b to master...
